File: src/Wc3HandleLua/Handle/Frame.lua
--=========
-- Include
--=========

local Class = LibManager.getDepency('LuaClass') or error('')
---@type Wc3Utils
local Wc3Utils = LibManager.getDepency('Wc3Utils') or error('')
local Color = Wc3Utils.Color or error('')
local isTypeErr = Wc3Utils.isTypeErr or error('')
local Log = Wc3Utils.Log or error('')

---@type HandleClass
local Handle = require('Handle') or error('')

--=======
-- Class
--=======

local Frame = Class.new('Frame', Handle)
---@class Frame : Handle
local public = Frame.public
---@class FrameClass : HandleClass
local static = Frame.static
---@type FrameClass
local override = Frame.override
local private = {}

--=========
-- Static
--=========

---@param name string
---@param is_simple boolean
---@param child Frame | nil
---@return Frame
function override.new(name, is_simple, child)
    isTypeErr(name, 'string', 'name')
    isTypeErr(is_simple, 'boolean', 'is_simple')
    if child then isTypeErr(child, Frame, 'child') end

    local handle = private.createFramehandle(name, is_simple)
    return static.link(handle, is_simple, child)
end

---@param handle framehandle
---@param is_simple boolean
---@param child Frame | nil
---@return Frame
function override.link(handle, is_simple, child)
    isTypeErr(handle, 'framehandle', 'handle')
    isTypeErr(is_simple, 'boolean', 'is_simple')
    if child then isTypeErr(child, Frame, 'child') end

    local instance = child or Class.allocate(Frame)
    instance = Handle.new(handle, BlzDestroyFrame, instance)
    private.newData(instance, is_simple)

    return instance
end

--========
-- Public
--========

---@return boolean
function public:isSimple()
    isTypeErr(self, Frame, 'self')
    return private.data[self].is_simple
end

---@param point framepointtype
---@param relative Frame
---@param relative_point framepointtype
---@param x number
---@param y number
function public:setPoint(point, relative, relative_point, x, y)
    isTypeErr(self, Frame, 'self')
    isTypeErr(point, 'framepoint', 'point')
    isTypeErr(relative, Frame, 'relative')
    isTypeErr(relative_point, 'framerelative_point', 'point')
    isTypeErr(x, 'number', 'x')
    isTypeErr(y, 'number', 'y')

    BlzFrameSetPoint(self:getData(), point,
                     relative:getData(), relative_point,
                     x, y)
end

---@param point framepointtype
---@param x number
---@param y number
function public:setAbsPoint(point, x, y)
    isTypeErr(self, Frame, 'self')
    isTypeErr(point, 'framepoint', 'point')
    isTypeErr(x, 'number', 'x')
    isTypeErr(y, 'number', 'y')

    BlzFrameSetAbsPoint(self:getData(), point, x, y)
end

function public:clearAllPoints()
    isTypeErr(self, Frame, 'self')
    BlzFrameClearAllPoints(self:getData())
end

---@param relative Frame
function public:setAllPoints(relative)
    isTypeErr(self, Frame, 'self')
    isTypeErr(relative, Frame, 'relative')
    BlzFrameSetAllPoints(self:getData(), relative:getData())
end

---@return boolean
function public:isVisible()
    isTypeErr(self, Frame, 'self')
    return BlzFrameIsVisible(self:getData())
end

---@param flag boolean
function public:setVisible(flag)
    isTypeErr(self, Frame, 'self')
    isTypeErr(flag, 'boolean', 'flag')

    BlzFrameSetVisible(self:getData(), flag)
end

---@return boolean
function public:isEnabled()
    isTypeErr(self, Frame, 'self')
    return BlzFrameGetEnable(self:getData())
end

---@param flag boolean
function public:setEnabled(flag)
    isTypeErr(self, Frame, 'self')
    isTypeErr(flag, 'boolean', 'flag')
    BlzFrameSetEnable(self:getData(), flag)
end

--- [0, 1]
---@param alpha number
function public:setAlpha(alpha)
    isTypeErr(self, Frame, 'self')
    isTypeErr(alpha, 'number', 'alpha')

    alpha = alpha > 1 and 1 or alpha < 0 and 0 or alpha
    BlzFrameSetAlpha(self:getData(), math.floor(255 * alpha))
end


---@param width number
---@param height number
function public:setSize(width, height)
    isTypeErr(self, Frame, 'self')
    isTypeErr(width, 'number', 'width')
    isTypeErr(height, 'number', 'height')
    BlzFrameSetSize(self:getData(), width, height)
end

---@return number
function public:getWidth()
    isTypeErr(self, Frame, 'self')
    return BlzFrameGetWidth(self:getData())
end

---@return number
function public:getHeight()
    isTypeErr(self, Frame, 'self')
    return BlzFrameGetHeight(self:getData())
end

---@param color Color
function public:setColor(color)
    isTypeErr(self, Frame, 'self')
    isTypeErr(color, Color, 'color')
    local priv = private.data[self]

    priv.color = Color.copy(color)
    local i_color = BlzConvertColor(math.floor(255 * color.r),
                                    math.floor(255 * color.g),
                                    math.floor(255 * color.b),
                                    math.floor(255 * color.a))
    BlzFrameSetVertexColor(self:getData(), i_color)
end

---@return Color
function public:getColor()
    isTypeErr(self, Frame, 'self')
    return Color.copy(private.data[self].color)
end

---@param level number
function public:setLevel(level)
    isTypeErr(self, Frame, 'self')
    isTypeErr(level, 'number', 'level')
    BlzFrameSetLevel(self:getData(), level)
end

---@param parent Frame | nil
function public:setParent(parent)
    isTypeErr(self, Frame, 'self')
    if parent then isTypeErr(parent, Frame, 'parent') end
    BlzFrameSetParent(self:getData(), parent and parent:getData() or nil)
end

---@return Frame | nil
function public:getParent()
    isTypeErr(self, Frame, 'self')
    return static.getLinked(BlzFrameGetParent(self:getData()))
end

---@param index number
---@return Frame | nil
function public:getChild(index)
    isTypeErr(self, Frame, 'self')
    isTypeErr(index, 'number', 'index')
    return static.getLinked(BlzFrameGetChild(self:getData(), index))
end

---@return number
function public:getChildrenCount()
    isTypeErr(self, Frame, 'self')
    return BlzFrameGetChildrenCount(self:getData())
end

--=========
-- Private
--=========

private.data = setmetatable({}, {__mode = 'k'})

---@param self Frame
---@param is_simple boolean
function private.newData(self, is_simple)
    local priv = {
        color = Color.new(1, 1, 1, 1),
        is_simple = is_simple,
    }
    private.data[self] = priv
end

---@param name string
---@param is_simple boolean
---@return framehandle
function private.createFramehandle(name, is_simple)
    local handle
    local test_handle
    if is_simple then
        handle = BlzCreateSimpleFrame(name, nil, 0)
        test_handle = BlzCreateSimpleFrame('', nil, 0)
    else
        handle = BlzCreateFrame(name, private.console_ui_backdrop, 0, 0)
        test_handle = BlzCreateFrame('', private.console_ui_backdrop, 0, 0)
    end

    if tostring(handle) == tostring(test_handle) then
        BlzDestroyFrame(handle)
        BlzDestroyFrame(test_handle)
        Log:err('Can not create framehandle', 3)
    end
    BlzDestroyFrame(test_handle)

    return handle
end

return static

File: src/Handle/Frame.ts
import { Vec2 } from '../Math'
import { Color } from "../Utils";
import { Handle } from "./Handle";

function createFramehandle(name: string, is_simple: boolean){
    let handle: jframehandle|undefined
    let test: jframehandle|undefined

    handle = is_simple ? BlzCreateSimpleFrame(name, null, 0) : BlzCreateFrame(name, null, 0, 0)
    test = is_simple ? BlzCreateSimpleFrame('', null, 0) : BlzCreateFrame('', null, 0, 0)

    if (tostring(handle) == tostring(test)){
        BlzDestroyFrame(handle)
        BlzDestroyFrame(test)
        throw('Can not create framehandle with name ' + name)
    }
    BlzDestroyFrame(test)
    return handle
}

export class hFrame extends Handle<jframehandle> {
    constructor(name: string, is_simple: boolean)
    constructor(handle: jframehandle, is_simple: boolean)
    constructor(name_or_handle: string|jframehandle, is_simple: boolean){
        let h
        if (typeof name_or_handle === 'string'){
            h = createFramehandle(name_or_handle, is_simple)
        } else {
            h = name_or_handle
        }

        super(h)
        this.__level = 0
        this.__color = new Color()
    }

    static get(id: jframehandle | number){
        return Handle.get(id, 'framehandle') as hFrame | undefined
    }

    get size(){
        return new Vec2(BlzFrameGetWidth(this.handle),
                        BlzFrameGetHeight(this.handle))
    }
    set size(size: Vec2){
        BlzFrameSetSize(this.handle, size.x, size.y)
    }

    get parent(){
        let h = BlzFrameGetParent(this.handle)
        return h ? hFrame.get(h) : undefined
    }
    set parent(parent: hFrame | undefined){
        BlzFrameSetParent(this.handle, parent ? parent.handle : undefined)
    }

    get visible(){return BlzFrameIsVisible(this.handle)};
    set visible(flag: boolean){
        BlzFrameSetVisible(this.handle, flag)
    }

    get enable(){return BlzFrameGetEnable(this.handle)}
    set enable(flag: boolean){
        BlzFrameSetEnable(this.handle, flag)
    }

    get level(){return this.__level}
    set level(lvl: number){
        this.__level = lvl
        BlzFrameSetLevel(this.handle, lvl)
    }

    get color(){return this.__color.copy()}
    set color(color: Color){
        this.__color = color.copy()
        BlzFrameSetVertexColor(this.handle, color.getWcCode())
    }

    setPoint(point: jframepointtype,
             other: hFrame, relative_point: jframepointtype,
             offset: Vec2){
        BlzFrameSetPoint(this.handle, point, other.handle, relative_point, offset.x, offset.y)
    }

    setPointAbs(point: jframepointtype, offset: Vec2){
        BlzFrameSetAbsPoint(this.handle, point, offset.x, offset.y)
    }

    setPointAll(other: hFrame){
        BlzFrameSetAllPoints(this.handle, other.handle)
    }

    freePointAll(){
        BlzFrameClearAllPoints(this.handle)
    }

    destroy(){
        BlzDestroyFrame(this.handle)
        super.destroy()
    }

    private __level: number
    private __color: Color
}